feat(Sticker): updates, sticker packs, and guild stickers

advaith1 opened the pull request:
**Please describe the changes this PR makes and why it should be merged:**

* Removes deprecated sticker.asset (which is just an empty string now)
* **Updates URL** (old one doesn't work anymore)
* Adds sticker properties: type, packID, available, guild, guildID, user, sortValue, fetchPack(), edit(), delete(), equals()
* Adds StickerPack
* Adds Client#fetchSticker(id)
* Adds Client#fetchNitroStickerPacks()
* Renames `MANAGE_EMOJIS` permission to `MANAGE_EMOJIS_AND_STICKERS` as requested by Jethro, to reflect client update
* Adds sending stickers (`stickers` in MessageOptions)
* Adds Guild#stickers and GuildStickerManager
* Adds `stickerCreate`, `stickerUpdate`, and `stickerDelete` events
* Renames `GUILD_EMOJIS` intent to `GUILD_EMOJIS_AND_STICKERS`
* Adds sticker audit log events

https://github.com/discord/discord-api-docs/pull/3127
https://github.com/discord/discord-api-docs/pull/3171
https://github.com/discord/discord-api-docs/pull/3128

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)
- This PR includes breaking changes (methods removed or renamed, parameters moved or removed)

advaith1 commented on src/structures/Sticker.js:
```diff
@@ -42,16 +42,40 @@ class Sticker extends Base {
-    this.packID = sticker.pack_id;
+    this.packID = sticker.pack_id ?? null;
 
     /**
      * An array of tags for the sticker, if any
      * @type {string[]}
      */
     this.tags = sticker.tags?.split(', ') ?? [];
```
I left `tags` as an array but for guild stickers this is just one word

advaith1 commented on src/structures/Sticker.js:
```diff
@@ -42,16 +46,40 @@ class Sticker extends Base {
+     */
+    this.guildID = sticker.guild_id ?? null;
+
+    /**
+     * The user that uploaded the guild sticker
+     * @type {?User}
+     */
+    this.user = sticker.user ? this.client.users.add(sticker.user) : null;
```
I left this as `user` like the API, but d.js uses `author` for emojis, should it be changed to `author` here too?

kyranet replied:
+1 for me, it's a bit misleading to use `sticker.user`. Although `author` can also be troublesome, because this field is about the user who uploaded it, not the author of the asset itself (see: people who re-upload emojis made by others from other servers). Unless, of course, we ignore that detail.

Depending on what we decide, we'll have to change [`GuildEmoji#author`](https://discord.js.org/#/docs/main/master/class/GuildEmoji?scrollTo=author) (in a later PR) to match.

cc: @iCrawl @SpaceEEC @vladfrangu

SpaceEEC replied:
+1 on `user`

vladfrangu replied:
+1 on `user`. Also, maybe consider adding a `false` to the `.add` method to not cache the user?

advaith1 replied:
> to not cache the user?

🤔 why? aren't users usually cached when received?

also, afaik this is only present when the sticker is fetched in the guild and the bot has perms, so this will currently always be null in d.js (unless the sticker object is manually constructed) since GuildStickerManager hasn't been implemented yet

vladfrangu replied:
Can you add a .fetch method then?

advaith1 replied:
ok looks like nobody responded in #library-discussion so I'll post it here:

![image](https://user-images.githubusercontent.com/11778454/123366203-a648d380-d52c-11eb-9604-240daa4400b3.png)

vladfrangu replied:
Then... let's do fetchAuthor, ala GuildEmoji? CC @iCrawl @kyranet

iCrawl replied:
Sure.

advaith1 commented:
ok, I think this is ready for review?

it doesn't add GuildStickerManager or the guild sticker update gw event bc those can't really be tested now, but I've tested everything added

kyranet commented on src/util/Constants.js:
```diff
@@ -169,6 +169,8 @@ exports.Endpoints = {
       AppAsset: (clientID, hash, { format = 'webp', size } = {}) =>
         makeImageUrl(`${root}/app-assets/${clientID}/${hash}`, { size, format }),
+      StickerBanner: (bannerID, format = 'webp', size) =>
+        makeImageUrl(`${root}/app-assets/710982414301790216/store/${bannerID}`, { size, format }),
```
What's with this oddly specific ID?

advaith1 replied:
all the sticker banners are on that "Stickers" application, its id doesn't seem to be received from the API so I assume that it's in the client code, I also included it in the docs pr

kyranet replied:
What about third-party stickers, such as Among Us's? Do they use the same application?

iCrawl replied:
![image](https://user-images.githubusercontent.com/20760160/122364191-765c6780-cf51-11eb-842f-2526a29320e8.png)

advaith1 replied:
hm i just noticed that the name for this is misleading: I put StickerBanner but it's more accurately StickerPackBanner, I'll fix that

guild stickers seem to not have packs so this wouldn't apply for them

kyranet replied:
. 　。　　　　•　 　ﾟ　　。 　　.　　　.　　　 　　
 .　　。　　 。　. 　.　　 。　 ඞ 。　 . • .
   @iCrawl was ejected. . . 　 。　. 　　。　　
　　　.　　　　　. ,　　　　.　 .　　 .

---

Are you able to test that just in case? @advaith1 

I need confirmation in this.

advaith1 replied:
I can confirm that guild stickers do not have `pack_id`, if that's what you mean?

kyranet replied:
Clarified from Discord - my main concern was about whether that ID is confirmed to be always constant (documented in Discord API's docs or an answer from a Discord developer in DDevs), for which @advaith1 has asked and is waiting for a reply.

I think it's prudent to wait for said question to be answered before we land this, as the hardcoded ID can change and potentially break this method in the future.

kyranet replied:
Ref: https://github.com/discord/discord-api-docs/pull/3128

advaith1 replied:
![image](https://user-images.githubusercontent.com/11778454/122484490-8ee87400-cf89-11eb-823c-570842751b16.jpeg)

kyranet commented on src/client/Client.js:
```diff
@@ -310,6 +312,36 @@ class Client extends BaseClient {
+   *   .then(sticker => console.log(`Obtained sticker with name: ${sticker.name}`))
+   *   .catch(console.error);
+   */
+  fetchSticker(id) {
+    return this.api
+      .stickers(id)
+      .get()
+      .then(data => new Sticker(this, data));
+  }
```
Please use async/await.

advaith1 replied:
none of the client methods around this use async/await though? was just being consistent with those

and if these did use async/await it would just be literally adding `async ` and `await ` right?

iCrawl replied:
It's not so much about consistency, but slowly converting to async/await. So newer methods should use async/await over thenables.

advaith1 replied:
updated, are they good now?

kyranet replied:
We usually prefer splitting those lines for the sake of readability, like in this case it'd be:

```javascript
const data = await this.api.stickers(id).get();
return new Sticker(this, data);
```

Inlining the call to save one variable is something I think we shouldn't do, specially when accessing to properties of it, as it adds a lot of parens.

kyranet commented on src/client/Client.js:
```diff
@@ -310,6 +312,36 @@ class Client extends BaseClient {
+   *   .then(packs => console.log(`Available sticker packs are: ${packs.map(pack => pack.name).join(', ')}`))
+   *   .catch(console.error);
+   */
+  fetchNitroStickerPacks() {
+    return this.api('sticker-packs')
+      .get()
+      .then(res => new Collection(res.sticker_packs.map(p => [p.id, new StickerPack(this, p)])));
+  }
```
Please use async/await.

kyranet commented on src/client/Client.js:
```diff
@@ -310,6 +312,36 @@ class Client extends BaseClient {
 
+  /**
+   * Obtains a sticker from Discord.
+   * @param {Snowflake} id ID of the sticker
```
```suggestion
   * @param {Snowflake} id The sticker's ID
```

kyranet reviewed (changes requested):
Just three small things, otherwise LGTM

kyranet reviewed (changes requested):
Just three small things, otherwise LGTM

SpaceEEC commented on src/client/Client.js:
```diff
@@ -310,6 +312,33 @@ class Client extends BaseClient {
+   *   .catch(console.error);
+   */
+  async fetchSticker(id) {
+    return new Sticker(this, await this.api.stickers(id).get());
```
What @kyranet said: https://github.com/discordjs/discord.js/pull/5867#discussion_r654795938

SpaceEEC commented on src/client/Client.js:
```diff
@@ -310,6 +312,33 @@ class Client extends BaseClient {
+   */
+  async fetchNitroStickerPacks() {
+    return new Collection(
+      (await this.api('sticker-packs').get()).sticker_packs.map(p => [p.id, new StickerPack(this, p)]),
```
What @kyranet said: https://github.com/discordjs/discord.js/pull/5867#discussion_r654795938

SpaceEEC reviewed (dismissed):
Otherwise looking good

ThaTiemsz commented on src/client/Client.js:
```diff
@@ -310,6 +312,33 @@ class Client extends BaseClient {
+   *   .then(packs => console.log(`Available sticker packs are: ${packs.map(pack => pack.name).join(', ')}`))
+   *   .catch(console.error);
+   */
+  async fetchNitroStickerPacks() {
```
```suggestion
  async fetchPremiumStickerPacks() {
```
Rename this perhaps for consistency? No other properties and methods use the term Nitro, they use premium.

vladfrangu commented on src/structures/Sticker.js:
```diff
@@ -72,16 +100,30 @@ class Sticker extends Base {
-    return `${this.client.options.http.cdn}/stickers/${this.id}/${this.asset}.${
-      this.format === 'LOTTIE' ? 'json' : 'png'
-    }`;
+    return `${this.client.options.http.cdn}/stickers/${this.id}.${this.format === 'LOTTIE' ? 'json' : 'png'}`;
```
Can't this be added to the CDN object in constants?

vladfrangu commented on src/structures/StickerPack.js:
```diff
@@ -0,0 +1,99 @@
+     * The ID of the sticker pack's banner image
+     * @type {?Snowflake}
+     */
+    this.bannerID = pack.banner_asset_id;
```
Is this property nullable? Can it be undefined?

advaith1 replied:
afaik this is always present and non-null, will ask for confirmation

advaith1 replied:
ok got confirmation that this is not optional or nullable, so resolving these

DTrombett replied:
If this is not nullable why the type is `?Snowflake`

vladfrangu commented on src/structures/StickerPack.js:
```diff
@@ -0,0 +1,99 @@
+  /**
+   * The URL to this sticker pack's banner.
+   * @param {StaticImageURLOptions} [options={}] Options for the Image URL
+   * @returns {?string}
```
If this is nullable, you should handle the no bannerID then

vladfrangu commented on typings/index.d.ts:
```diff
@@ -1755,6 +1764,41 @@ declare module 'discord.js' {
+    public type: StickerType;
+    public user: ?User;
+    public readonly url: string;
+    public fetchPack(): Promise<?StickerPack>;
```
This should be `StickerPack | null`

ThaTiemsz replied:
```suggestion
    public fetchPack(): Promise<StickerPack | null>;
```

vladfrangu commented on typings/index.d.ts:
```diff
@@ -1755,6 +1764,41 @@ declare module 'discord.js' {
+    public description: string;
+    public format: StickerFormatType;
+    public readonly guild: ?Guild;
+    public guildID: ?Snowflake;
```
All nullable properties should be `actualType | null`

ThaTiemsz replied:
```suggestion
    public guildID: Snowflake | null;
```

NotSugden commented on src/client/Client.js:
```diff
@@ -310,6 +312,33 @@ class Client extends BaseClient {
+   */
+  async fetchNitroStickerPacks() {
+    const data = await this.api('sticker-packs').get();
+    return new Collection(data.sticker_packs.map(p => [p.id, new StickerPack(this, p)]));
```
perhaps use reduce? just to be more consistent

advaith1 replied:
imo this is cleaner than reduce, and Vlad said that he thinks it's fine

maybe the ones that use reduce to fill a collection can be changed to map instead

cc @iCrawl @kyranet

kyranet replied:
I don't recall if this was one of the patterns we switched from when moving towards reduce, or if it was never written like that.

Either way, I'm fine with this pattern.

NotSugden commented on src/structures/StickerPack.js:
```diff
@@ -0,0 +1,99 @@
+     * The stickers in the pack
+     * @type {Collection<Snowflake, Sticker>}
+     */
+    this.stickers = new Collection(pack.stickers.map(s => [s.id, new Sticker(client, s)]));
```
same as previous

NotSugden commented on src/structures/Sticker.js:
```diff
@@ -1,14 +1,18 @@
 class Sticker extends Base {
+  /**
+   * @param {Client} client The instantiating client
+   * @param {Object} sticker The data for the sticker
```
should there be a jsdoc type for the sticker data? (like what was introduced in #5862)

NotSugden commented on src/structures/StickerPack.js:
```diff
@@ -0,0 +1,99 @@
+class StickerPack extends Base {
+  /**
+   * @param {Client} client The instantiating client
+   * @param {Object} pack The data for the sticker pack
```
same here

ThaTiemsz commented on typings/index.d.ts:
```diff
@@ -1755,6 +1765,43 @@ declare module 'discord.js' {
+    constructor(client: Client, data: unknown);
+    public readonly createdTimestamp: number;
+    public readonly createdAt: Date;
+    public available: ?boolean;
```
```suggestion
    public available: boolean | null;
```

advaith1 replied:
already discussed with Vlad, both work

ThaTiemsz replied:
@advaith1 It's better for consistency since `| null` is used everywhere else in the typings. Also using the `?` prefix will cause it to fail to compile:

![image](https://user-images.githubusercontent.com/18150845/124066490-082ba080-da39-11eb-95d7-8379a470178c.png)

ThaTiemsz commented on typings/index.d.ts:
```diff
@@ -1755,6 +1765,43 @@ declare module 'discord.js' {
+    public readonly createdTimestamp: number;
+    public readonly createdAt: Date;
+    public available: ?boolean;
+    public description: ?string;
```
```suggestion
    public description: string | null;
```

ThaTiemsz commented on typings/index.d.ts:
```diff
@@ -1755,6 +1765,43 @@ declare module 'discord.js' {
+    public available: ?boolean;
+    public description: ?string;
+    public format: StickerFormatType;
+    public readonly guild: ?Guild;
```
```suggestion
    public readonly guild: Guild | null;
```

ThaTiemsz commented on typings/index.d.ts:
```diff
@@ -1755,6 +1765,43 @@ declare module 'discord.js' {
+    public id: Snowflake;
+    public name: string;
+    public packID: ?Snowflake;
+    public sortValue: ?number;
```
```suggestion
    public sortValue: number | null;
```

ThaTiemsz commented on typings/index.d.ts:
```diff
@@ -1755,6 +1765,43 @@ declare module 'discord.js' {
+    public guildID: ?Snowflake;
+    public id: Snowflake;
+    public name: string;
+    public packID: ?Snowflake;
```
```suggestion
    public packID: Snowflake | null;
```

ThaTiemsz commented on typings/index.d.ts:
```diff
@@ -1755,6 +1765,43 @@ declare module 'discord.js' {
+    public name: string;
+    public packID: ?Snowflake;
+    public sortValue: ?number;
+    public tags: ?string[];
```
```suggestion
    public tags: string[] | null;
```

ThaTiemsz commented on typings/index.d.ts:
```diff
@@ -1755,6 +1765,43 @@ declare module 'discord.js' {
+    public packID: ?Snowflake;
+    public sortValue: ?number;
+    public tags: ?string[];
+    public type: ?StickerType;
```
```suggestion
    public type: StickerType | null;
```

ThaTiemsz commented on typings/index.d.ts:
```diff
@@ -1755,6 +1765,43 @@ declare module 'discord.js' {
+    public sortValue: ?number;
+    public tags: ?string[];
+    public type: ?StickerType;
+    public user: ?User;
```
```suggestion
    public user: User | null;
```

ThaTiemsz commented on typings/index.d.ts:
```diff
@@ -1755,6 +1765,43 @@ declare module 'discord.js' {
+    public type: ?StickerType;
+    public user: ?User;
+    public readonly url: string;
+    public fetch(): Promise<?Sticker>;
```
```suggestion
    public fetch(): Promise<Sticker | null>;
```

ThaTiemsz commented on typings/index.d.ts:
```diff
@@ -1755,6 +1765,43 @@ declare module 'discord.js' {
+    public readonly url: string;
+    public fetch(): Promise<?Sticker>;
+    public fetchPack(): Promise<?StickerPack>;
+    public fetchUser(): Promise<?User>;
```
```suggestion
    public fetchUser(): Promise<User | null>;
```

ThaTiemsz commented on typings/index.d.ts:
```diff
@@ -1755,6 +1765,43 @@ declare module 'discord.js' {
+    public readonly createdTimestamp: number;
+    public readonly createdAt: Date;
+    public bannerID: Snowflake;
+    public readonly coverSticker: ?Sticker;
```
```suggestion
    public readonly coverSticker: Sticker | null;
```

ThaTiemsz commented on typings/index.d.ts:
```diff
@@ -1755,6 +1765,43 @@ declare module 'discord.js' {
+    public readonly createdAt: Date;
+    public bannerID: Snowflake;
+    public readonly coverSticker: ?Sticker;
+    public coverStickerID: ?Snowflake;
```
```suggestion
    public coverStickerID: Snowflake | null;
```

vladfrangu commented on src/errors/Messages.js:
```diff
@@ -98,8 +98,9 @@ const Messages = {
-  MISSING_MANAGE_EMOJIS_PERMISSION: guild =>
-    `Client must have Manage Emoji permission in guild ${guild} to see emoji authors.`,
+  MISSING_MANAGE_EMOJIS_AND_STICKERS_PERMISSION: guild =>
+    `Client must have Manage Emojis and Stickers permission in guild ${guild} to see emoji authors.`,
```
```suggestion
    `Client must have Manage Emojis and Stickers permission in guild ${guild} to see the sticker author.`,
```

advaith1 replied:
this error is only used by GuildEmoji#fetchAuthor, I just renamed it bc the permission was renamed

vladfrangu replied:
Ah, gotcha!

vladfrangu commented on src/errors/Messages.js:
```diff
@@ -98,8 +98,9 @@ const Messages = {
-    `Client must have Manage Emoji permission in guild ${guild} to see emoji authors.`,
+  MISSING_MANAGE_EMOJIS_AND_STICKERS_PERMISSION: guild =>
+    `Client must have Manage Emojis and Stickers permission in guild ${guild} to see emoji authors.`,
+  NOT_GUILD_STICKER: 'Sticker is a standard (non-guild) sticker and has no User.',
```
```suggestion
  NOT_GUILD_STICKER: 'Sticker is a standard (non-guild) sticker and has no author.',
```

vladfrangu commented on src/structures/Sticker.js:
```diff
@@ -72,16 +105,62 @@ class Sticker extends Base {
+
+  /**
+   * Fetches this sticker.
+   * @returns {Promise<?Sticker>}
```
```suggestion
   * @returns {Promise<Sticker>}
```

vladfrangu commented on src/structures/Sticker.js:
```diff
@@ -72,16 +105,62 @@ class Sticker extends Base {
+   * @returns {Promise<?StickerPack>}
+   */
+  async fetchPack() {
+    return (this.packID && (await this.client.fetchPremiumStickerPacks()).get(this.packID)) ?? null;
```
```suggestion
    return this.packID ? (await this.client.fetchPremiumStickerPacks()).get(this.packID) ?? null : null;
```

NotSugden replied:
whats wrong with `this.packID && (...)`?

vladfrangu replied:
I never said something's wrong with it, but I don't know if that's the style we wanna go with everywhere... @iCrawl @kyranet thoughts?

iCrawl replied:
Generally I would say this is more readable:
`(this.packID && (await this.client.fetchPremiumStickerPacks()).get(this.packID)) ?? null;`
than this:
`this.packID ? (await this.client.fetchPremiumStickerPacks()).get(this.packID) ?? null : null;`

kyranet replied:
I prefer the current code's approach.

vladfrangu commented on src/structures/Sticker.js:
```diff
@@ -72,16 +105,62 @@ class Sticker extends Base {
+   * @readonly
+   */
+  get partial() {
+    return 'username' in this;
```
Wait hold up, this doesn't exist on a sticker

advaith1 replied:
oops

advaith1 replied:
fixed

BannerBomb commented on src/structures/Sticker.js:
```diff
@@ -72,16 +105,62 @@ class Sticker extends Base {
+   * @readonly
+   */
+  get partial() {
+    return 'type' in this;
```
Shouldn't this be 
```suggestion
    return !('type' in this);
```

BannerBomb replied:
also not sure if it was intended but partial will always be false since the `type` property is always being added to the class.

advaith1 replied:
yeah i'll redo this

BannerBomb commented on src/structures/Sticker.js:
```diff
@@ -42,16 +51,40 @@ class Sticker extends Base {
     /**
-     * An array of tags for the sticker, if any
+     * An array of tags for the sticker
      * @type {string[]}
```
```suggestion
     * @type {?string[]}
```

BannerBomb commented on typings/index.d.ts:
```diff
@@ -1755,6 +1765,43 @@ declare module 'discord.js' {
+    public readonly url: string;
+    public fetch(): Promise<Sticker>;
+    public fetchPack(): Promise<StickerPack | null>;
+    public fetchUser(): Promise<User | null>;
```
```suggestion
    public fetchUser(): Promise<User | null>;
    public readonly partial: boolean;
```

advaith1 commented:
* added GuildStickerManager (Guild#stickers)
* added some methods to Stickers
* added `stickerCreate`, `stickerUpdate`, `stickerDelete`
* renamed `GUILD_EMOJIS` intent to `GUILD_EMOJIS_AND_STICKERS`

had to update APIRequest.json to get sticker creation working:
- file.key will override file.name as the key in the FormData body, but file.name will stay the actual filename
- options.dontUsePayloadJSON makes it send each data field separately in the body, instead of in one `payload_json` field

these are both backwards compatible

will update typings tomorrow

iCrawl commented on src/client/Client.js:
```diff
@@ -303,6 +305,33 @@ class Client extends BaseClient {
 
+  /**
+   * Obtains a sticker from Discord.
+   * @param {Snowflake} id The sticker's ID
```
```suggestion
   * @param {Snowflake} id The sticker's id
```

iCrawl commented on src/managers/GuildStickerManager.js:
```diff
@@ -0,0 +1,160 @@
+class GuildStickerManager extends BaseManager {
+  constructor(guild, iterable) {
+    super(guild.client, iterable, Sticker);
+    /**
```
```suggestion

    /**
```

iCrawl commented on src/managers/GuildStickerManager.js:
```diff
@@ -0,0 +1,160 @@
+
+  /**
+   * Resolves an StickerResolvable to an Sticker ID string.
+   * @method resolveID
```
```suggestion
   * @method resolveId
```

iCrawl commented on src/managers/GuildStickerManager.js:
```diff
@@ -0,0 +1,160 @@
+   */
+
+  /**
+   * Resolves an StickerResolvable to an Sticker ID string.
```
```suggestion
   * Resolves an StickerResolvable to an Sticker id string.
```

iCrawl commented on src/managers/GuildStickerManager.js:
```diff
@@ -0,0 +1,160 @@
+   * @returns {Promise<Sticker>}
+   */
+  async edit(sticker, data, reason) {
+    const stickerID = this.resolveID(sticker);
```
```suggestion
    const stickerId = this.resolveId(sticker);
```

iCrawl commented on src/managers/GuildStickerManager.js:
```diff
@@ -0,0 +1,160 @@
+    const stickerID = this.resolveID(sticker);
+    if (!stickerID) throw new TypeError('INVALID_TYPE', 'sticker', 'StickerResolvable');
+
+    const d = await this.client.api.guilds(this.guild.id).stickers(stickerID).patch({
```
```suggestion
    const d = await this.client.api.guilds(this.guild.id).stickers(stickerId).patch({
```

iCrawl commented on src/managers/GuildStickerManager.js:
```diff
@@ -0,0 +1,160 @@
+      reason,
+    });
+
+    const existing = this.cache.get(stickerID);
```
```suggestion
    const existing = this.cache.get(stickerId);
```

iCrawl commented on src/managers/GuildStickerManager.js:
```diff
@@ -0,0 +1,160 @@
+   * @returns {Promise<void>}
+   */
+  async delete(sticker, reason) {
+    sticker = this.resolveID(sticker);
```
```suggestion
    sticker = this.resolveId(sticker);
```

iCrawl commented on src/managers/GuildStickerManager.js:
```diff
@@ -0,0 +1,160 @@
+
+  /**
+   * Obtains one or more stickers from Discord, or the sticker cache if they're already available.
+   * @param {Snowflake} [id] ID of the sticker
```
```suggestion
   * @param {Snowflake} [id] The Sticker's id
```

iCrawl commented on src/structures/Sticker.js:
```diff
@@ -1,33 +1,42 @@
+
+  _patch(sticker) {
     /**
      * The ID of the sticker
```
```suggestion
     * The Sticker's id
```

iCrawl commented on src/structures/Sticker.js:
```diff
@@ -42,16 +51,40 @@ class Sticker extends Base {
     /**
-     * The ID of the pack the sticker is from
-     * @type {Snowflake}
+     * The ID of the pack the sticker is from, for standard stickers
```
```suggestion
     * The id of the pack the sticker is from, for standard stickers
```

iCrawl commented on src/structures/Sticker.js:
```diff
@@ -42,16 +51,40 @@ class Sticker extends Base {
+     * The ID of the pack the sticker is from, for standard stickers
+     * @type {?Snowflake}
+     */
+    this.packID = sticker.pack_id ?? null;
```
```suggestion
    this.packId = sticker.pack_id ?? null;
```

iCrawl commented on src/structures/Sticker.js:
```diff
@@ -42,16 +51,40 @@ class Sticker extends Base {
+     * @type {?Snowflake}
      */
-    this.packID = sticker.pack_id;
+    this.guildID = sticker.guild_id ?? null;
```
```suggestion
    this.guildId = sticker.guild_id ?? null;
```

iCrawl commented on src/structures/Sticker.js:
```diff
@@ -42,16 +51,40 @@ class Sticker extends Base {
+    this.available = sticker.available ?? null;
+
+    /**
+     * The ID of the guild that owns this sticker
```
```suggestion
     * The id of the guild that owns this sticker
```

iCrawl commented on src/structures/Sticker.js:
```diff
@@ -72,16 +105,125 @@ class Sticker extends Base {
+   * @readonly
+   */
+  get guild() {
+    return this.client.guilds.resolve(this.guildID);
```
```suggestion
    return this.client.guilds.resolve(this.guildId);
```

iCrawl commented on src/structures/Sticker.js:
```diff
@@ -72,16 +105,125 @@ class Sticker extends Base {
+   * @returns {Promise<?StickerPack>}
+   */
+  async fetchPack() {
+    return (this.packID && (await this.client.fetchPremiumStickerPacks()).get(this.packID)) ?? null;
```
```suggestion
    return (this.packId && (await this.client.fetchPremiumStickerPacks()).get(this.packId)) ?? null;
```

iCrawl commented on src/structures/Sticker.js:
```diff
@@ -72,16 +105,125 @@ class Sticker extends Base {
+    if (this.partial) await this.fetch();
+    if (!this.guildID) throw new Error('NOT_GUILD_STICKER');
+
+    const data = await this.client.api.guilds(this.guildID).stickers(this.id).get();
```
```suggestion
    const data = await this.client.api.guilds(this.guildId).stickers(this.id).get();
```

iCrawl commented on src/structures/Sticker.js:
```diff
@@ -72,16 +105,125 @@ class Sticker extends Base {
+        other.tags.length === this.tags.length &&
+        other.tags.every(tag => this.tags.includes(tag)) &&
+        other.available === this.available &&
+        other.guildID === this.guildID &&
```
```suggestion
        other.guildId === this.guildId &&
```

iCrawl commented on src/structures/Sticker.js:
```diff
@@ -72,16 +105,125 @@ class Sticker extends Base {
+        other.type === this.type &&
+        other.format === this.format &&
+        other.name === this.name &&
+        other.packID === this.packID &&
```
```suggestion
        other.packId === this.packId &&
```

iCrawl commented on src/structures/StickerPack.js:
```diff
@@ -0,0 +1,99 @@
+     * The ID of the pack's SKU
+     * @type {Snowflake}
+     */
+    this.skuID = pack.sku_id;
```
```suggestion
    this.skuId = pack.sku_id;
```

iCrawl commented on src/structures/StickerPack.js:
```diff
@@ -0,0 +1,99 @@
+    this.name = pack.name;
+
+    /**
+     * The ID of the pack's SKU
```
```suggestion
     * The id of the pack's SKU
```

iCrawl commented on src/structures/StickerPack.js:
```diff
@@ -0,0 +1,99 @@
+  constructor(client, pack) {
+    super(client);
+    /**
+     * The ID of the sticker pack
```
```suggestion
     * The Sticker pack's id
```

iCrawl commented on src/structures/StickerPack.js:
```diff
@@ -0,0 +1,99 @@
+     * The ID of a sticker in the pack which is shown as the pack's icon
+     * @type {?Snowflake}
+     */
+    this.coverStickerID = pack.cover_sticker_id ?? null;
```
```suggestion
    this.coverStickerId = pack.cover_sticker_id ?? null;
```

iCrawl commented on src/structures/StickerPack.js:
```diff
@@ -0,0 +1,99 @@
+     * The ID of the sticker pack's banner image
+     * @type {Snowflake}
+     */
+    this.bannerID = pack.banner_asset_id;
```
```suggestion
    this.bannerId = pack.banner_asset_id;
```

iCrawl commented on src/structures/StickerPack.js:
```diff
@@ -0,0 +1,99 @@
+    this.description = pack.description;
+
+    /**
+     * The ID of the sticker pack's banner image
```
```suggestion
     * The id of the sticker pack's banner image
```

iCrawl commented on src/structures/StickerPack.js:
```diff
@@ -0,0 +1,99 @@
+    this.skuID = pack.sku_id;
+
+    /**
+     * The ID of a sticker in the pack which is shown as the pack's icon
```
```suggestion
     * The id of a sticker in the pack which is shown as the pack's icon
```

iCrawl commented on src/structures/StickerPack.js:
```diff
@@ -0,0 +1,99 @@
+   * @readonly
+   */
+  get coverSticker() {
+    return this.coverStickerID && this.stickers.get(this.coverStickerID);
```
```suggestion
    return this.coverStickerId && this.stickers.get(this.coverStickerId);
```

iCrawl commented on src/structures/StickerPack.js:
```diff
@@ -0,0 +1,99 @@
+   * @returns {string}
+   */
+  bannerURL({ format, size } = {}) {
+    return this.client.rest.cdn.StickerPackBanner(this.bannerID, format, size);
```
```suggestion
    return this.client.rest.cdn.StickerPackBanner(this.bannerId, format, size);
```

iCrawl commented on src/util/Constants.js:
```diff
@@ -170,6 +170,8 @@ exports.Endpoints = {
         makeImageUrl(`${root}/app-icons/${clientID}/${hash}`, { size, format }),
       AppAsset: (clientID, hash, { format = 'webp', size } = {}) =>
         makeImageUrl(`${root}/app-assets/${clientID}/${hash}`, { size, format }),
+      StickerPackBanner: (bannerID, format = 'webp', size) =>
```
```suggestion
      StickerPackBanner: (bannerId, format = 'webp', size) =>
```

iCrawl commented on src/util/Constants.js:
```diff
@@ -170,6 +170,8 @@ exports.Endpoints = {
       AppAsset: (clientID, hash, { format = 'webp', size } = {}) =>
         makeImageUrl(`${root}/app-assets/${clientID}/${hash}`, { size, format }),
+      StickerPackBanner: (bannerID, format = 'webp', size) =>
+        makeImageUrl(`${root}/app-assets/710982414301790216/store/${bannerID}`, { size, format }),
```
```suggestion
        makeImageUrl(`${root}/app-assets/710982414301790216/store/${bannerId}`, { size, format }),
```

iCrawl commented on src/util/Constants.js:
```diff
@@ -178,6 +180,8 @@ exports.Endpoints = {
         makeImageUrl(`${root}/discovery-splashes/${guildID}/${hash}`, { size, format }),
       TeamIcon: (teamID, hash, { format = 'webp', size } = {}) =>
         makeImageUrl(`${root}/team-icons/${teamID}/${hash}`, { size, format }),
+      Sticker: (stickerID, stickerFormat) =>
```
```suggestion
      Sticker: (stickerId, stickerFormat) =>
```

iCrawl commented on src/util/Constants.js:
```diff
@@ -178,6 +180,8 @@ exports.Endpoints = {
       TeamIcon: (teamID, hash, { format = 'webp', size } = {}) =>
         makeImageUrl(`${root}/team-icons/${teamID}/${hash}`, { size, format }),
+      Sticker: (stickerID, stickerFormat) =>
+        `${root}/stickers/${stickerID}.${stickerFormat === 'LOTTIE' ? 'json' : 'png'}`,
```
```suggestion
        `${root}/stickers/${stickerId}.${stickerFormat === 'LOTTIE' ? 'json' : 'png'}`,
```

iCrawl commented on typings/index.d.ts:
```diff
@@ -630,6 +637,7 @@ declare module 'discord.js' {
         ) => string;
         AppIcon: (userID: Snowflake | number, hash: string, format: AllowedImageFormat, size: number) => string;
         AppAsset: (userID: Snowflake | number, hash: string, format: AllowedImageFormat, size: number) => string;
+        StickerPackBanner: (bannerID: Snowflake, format: AllowedImageFormat, size: number) => string;
```
```suggestion
        StickerPackBanner: (bannerId: Snowflake, format: AllowedImageFormat, size: number) => string;
```

iCrawl commented on typings/index.d.ts:
```diff
@@ -639,6 +647,7 @@ declare module 'discord.js' {
           size: number,
         ) => string;
         TeamIcon: (teamID: Snowflake | number, hash: string, format: AllowedImageFormat, size: number) => string;
+        Sticker: (stickerID: Snowflake, stickerFormat: StickerFormatType) => string;
```
```suggestion
        Sticker: (stickerId: Snowflake, stickerFormat: StickerFormatType) => string;
```

iCrawl commented on typings/index.d.ts:
```diff
@@ -1898,6 +1908,44 @@ declare module 'discord.js' {
+    public description: string | null;
+    public format: StickerFormatType;
+    public readonly guild: Guild | null;
+    public guildID: Snowflake | null;
```
```suggestion
    public guildId: Snowflake | null;
```

iCrawl commented on typings/index.d.ts:
```diff
@@ -1898,6 +1908,44 @@ declare module 'discord.js' {
+    public guildID: Snowflake | null;
+    public id: Snowflake;
+    public name: string;
+    public packID: Snowflake | null;
```
```suggestion
    public packId: Snowflake | null;
```

iCrawl commented on typings/index.d.ts:
```diff
@@ -1898,6 +1908,44 @@ declare module 'discord.js' {
+    constructor(client: Client, data: unknown);
+    public readonly createdTimestamp: number;
+    public readonly createdAt: Date;
+    public bannerID: Snowflake;
```
```suggestion
    public bannerId: Snowflake;
```

iCrawl commented on typings/index.d.ts:
```diff
@@ -1898,6 +1908,44 @@ declare module 'discord.js' {
+    public readonly createdAt: Date;
+    public bannerID: Snowflake;
+    public readonly coverSticker: Sticker | null;
+    public coverStickerID: Snowflake | null;
```
```suggestion
    public coverStickerId: Snowflake | null;
```

iCrawl commented on typings/index.d.ts:
```diff
@@ -1898,6 +1908,44 @@ declare module 'discord.js' {
+    public description: string;
+    public id: Snowflake;
+    public name: string;
+    public skuID: Snowflake;
```
```suggestion
    public skuId: Snowflake;
```

iCrawl reviewed (changes requested):
`ID` -> `Id` changes required.

iCrawl commented:
Additionally this needs a rebase.

iCrawl commented:
This needs a rebase for the typings 👍

almeidx commented on typings/index.d.ts:
```diff
@@ -1865,6 +1879,47 @@ declare module 'discord.js' {
+    public edit(data?: GuildStickerEditData, reason?: string): Promise<Sticker>;
+    public delete(reason?: string): Promise<Sticker>;
+    public equals(other: Sticker | unknown): boolean;
+    public readonly partial: boolean;
```
The `partial` property should be below `packId`

iCrawl commented:
This needs a rebase.

vladfrangu commented on src/rest/APIRequest.js:
```diff
@@ -43,8 +43,16 @@ class APIRequest {
+      }
+      if (typeof this.options.data !== 'undefined') {
+        if (this.options.dontUsePayloadJSON) {
+          for (const key in this.options.data) body.append(key, this.options.data[key]);
```
Please use for-of and Object.entries

vladfrangu reviewed (approved):
This needs a rebase

NotSugden commented on src/managers/GuildStickerManager.js:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    const data = await this.client.api.guilds(this.guild.id).stickers.get();
+    const stickers = new Collection(data.map(sticker => [sticker.id, this.add(sticker, cache)]));
```
```suggestion
    return new Collection(data.map(sticker => [sticker.id, this.add(sticker, cache)]));
```
can't multiline comment on mobile, line below needs to be removed also

NotSugden commented on src/rest/APIRequest.js:
```diff
@@ -43,8 +43,16 @@ class APIRequest {
-      for (const file of this.options.files) if (file && file.file) body.append(file.name, file.file, file.name);
-      if (typeof this.options.data !== 'undefined') body.append('payload_json', JSON.stringify(this.options.data));
+      for (const file of this.options.files) {
+        if (file && file.file) body.append(file.key ?? file.name, file.file, file.name);
```
```suggestion
        if (file?.file) body.append(file.key ?? file.name, file.file, file.name);
```

iCrawl commented:
This needs a rebase.

ITZProGamerDieYT commented:
you aware changes weren't merged?

kyranet commented:
Upstream PR has landed, and this needs a rebase 😄
